utils/plugins/utils.go

```diff
@@ -0,0 +1,202 @@
+	if err != nil {
+		return err
+	}
+	if plugins.Version != coreutils.GetPluginsYamlVersion() {
+		return errorutils.CheckError(errors.New(fmt.Sprintf("Expected plugins version in 'plugins.yaml is %d  but the actual value is %d", coreutils.GetPluginsYamlVersion(), plugins.Version)))
+	}
+	return nil
+}
+
+func readPluginsYaml() (*PluginsV1, error) {
+	plugins := new(PluginsV1)
+	content, err := getPluginsYamlFile()
+	if err != nil {
+		return nil, err
+	}
+	if len(content) == 0 {
+		// No plugins.yaml file was found, that means that we are in v0.
+		// Convert plugins layout to the latest version.
+		return convertPluginsV0ToV1()
+	}
+
+	err = json.Unmarshal(content, &plugins)
+	if err != nil {
+		return nil, errorutils.CheckError(err)
+	}
+	return plugins, err
+}
+
+func getPluginsYamlFile() (content []byte, err error) {
+	pluginsFilePath, err := getPluginsFilePath()
+	if err != nil {
+		return
+	}
+	exists, err := fileutils.IsFileExists(pluginsFilePath, false)
+	if err != nil {
+		return
+	}
+	if exists {
+		content, err = fileutils.ReadFile(pluginsFilePath)
+	}
+	return
+}
+
+// V0: in plugins directory there was no 'plugins.yaml' file. all executable files were in the same directory.
+// V1: create 'plugins.yaml' file inside 'plugins' directory. change the file's hierarchy inside 'plugins' directory.
+func convertPluginsV0ToV1() (*PluginsV1, error) {
+	err := migrateFileSystemLayoutV0ToV1()
+	if err != nil {
+		return nil, err
+	}
+	return CreatePluginsYamlFile()
+}
+
+// Change the file's hierarchy inside 'plugins' directory to:
+//	plugins (dir)
+//		plugin-name (dir)
+//			bin (dir)
+//				plugin-executable (file)
+//			resources:(optional dir)
+//				... (directories/files)
+func migrateFileSystemLayoutV0ToV1() error {
+	plugins, err := coreutils.GetPluginsDirectoryContent()
+	if err != nil {
+		return err
+	}
+	pluginsDir, err := coreutils.GetJfrogPluginsDir()
+	for _, p := range plugins {
+		// Skip 'plugins.yaml'
+		if p.Name() == coreutils.JfrogPluginsFile {
+			continue
+		}
+		if p.IsDir() {
+			log.Error("unexpected directory in plugins directory")
+			break
+		}
+		// Verify that the file is an executable file
+		if !IsExec(p.Mode()) {
+			log.Error("unexpected file in plugins directory: " + p.Name())
+			continue
+		}
+		// Move plugins exec files inside a directory with the plugin's name.
+		// Create a directory with the plugin's name + "_dir" extension, move the file inside and change directory's name back to plugin's name only.
+		pluginsName := getPluginsNameFromExec(p.Name())
+		err = os.MkdirAll(filepath.Join(pluginsDir, pluginsName+"_dir", coreutils.PluginsExecDirName), 0777)
+		if err != nil {
+			return err
+		}
+		err = fileutils.MoveFile(filepath.Join(pluginsDir, p.Name()), filepath.Join(pluginsDir, pluginsName+"_dir", coreutils.PluginsExecDirName, p.Name()))
+		if err != nil {
+			return err
+		}
+		err = os.Rename(filepath.Join(pluginsDir, pluginsName+"_dir"), filepath.Join(pluginsDir, pluginsName))
```

Comment:
We're noy allowed to use the ```os.Rename``` function, because ti does not work inside docker containers. See more info in the documentation of the ```fileutils.MoveFile``` function.